File: app/src/main/java/com/sk/revisit/managers/WebStorageManager.java
package com.sk.revisit.managers;

import android.net.Uri;
import android.webkit.URLUtil;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.sk.revisit.MyUtils;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;

import okhttp3.Headers;

public class WebStorageManager {
	private static final String TAG = "WebStorageManager";
	private static final String GET_METHOD = "GET";
	private static final String UTF_8 = "UTF-8";
	private static final String NO_OFFLINE_FILE_MESSAGE = "No offline file available.";
	private final MyUtils utils;

	public WebStorageManager(MyUtils utils) {
		this.utils = utils;
	}

	@Nullable
	public WebResourceResponse getResponse(@NonNull WebResourceRequest request) {
		MyUtils.requests.incrementAndGet();

		if (!GET_METHOD.equals(request.getMethod())) {
			utils.log(TAG, "Request method is not GET: " + request.getMethod());
			return null;
		}

		Uri uri = request.getUrl();
		String uriStr = uri.toString();
		utils.saveUrl(uriStr);

		if (!URLUtil.isNetworkUrl(uriStr)) {
			utils.log(TAG, "Not a network URL: " + uriStr);
			return null;
		}

		String localPath = utils.buildLocalPath(uri);
		if (localPath == null) {
			utils.log(TAG, "Could not build local path for: " + uriStr);
			return null;
		}

		File localFile = new File(localPath);
		if (localFile.exists()) {
			if (MyUtils.shouldUpdate && MyUtils.isNetworkAvailable) {
				utils.download(uri, createDownloadListener(uriStr, localPath));
			}
			return loadFromLocal(localFile, uri);
		} else {
			if (MyUtils.isNetworkAvailable) {
				utils.download(uri, createDownloadListener(uriStr, localPath));
				return loadFromLocal(localFile, uri);
			}
			utils.saveReq(uriStr);
			return createNoOfflineFileResponse();
		}
	}

	@NonNull
	private MyUtils.DownloadListener createDownloadListener(String uriStr, String localPath) {
		return new MyUtils.DownloadListener() {
			@Override
			public void onSuccess(File file, Headers headers) {
				MyUtils.resolved.incrementAndGet();
				utils.saveResp(String.format("[\"%s\",\"%s\",%d,\"%s\"]", uriStr, localPath, file.length(), headers.toString()));
			}

			@Override
			public void onFailure(Exception e) {
				MyUtils.failed.incrementAndGet();
				utils.saveReq(uriStr);
				utils.log(TAG, "Download failed for: " + uriStr, e);
			}
		};
	}

	@Nullable
	private WebResourceResponse loadFromLocal(@NonNull File localFile, Uri uri) {
		String localFilePath = localFile.getAbsolutePath();
		try {
			String mimeType = getMimeType(localFilePath, uri);
			MyUtils.resolved.incrementAndGet();
			InputStream inputStream = new FileInputStream(localFile);
			return new WebResourceResponse(mimeType, UTF_8, inputStream);
		} catch (FileNotFoundException e) {
			MyUtils.failed.incrementAndGet();
			utils.log(TAG, "File not found: " + localFile.getAbsolutePath(), e);
			return null;
		}
	}

	@NonNull
	private String getMimeType(String localFilePath, Uri uri) {
		String mimeType = utils.getMimeTypeFromMeta(localFilePath);
		if (mimeType == null) {
			utils.createMimeTypeMeta(uri);
			mimeType = utils.getMimeType(localFilePath);
		}
		return mimeType;
	}

	@NonNull
	private WebResourceResponse createNoOfflineFileResponse() {
		return new WebResourceResponse("text/html", UTF_8, new ByteArrayInputStream(NO_OFFLINE_FILE_MESSAGE.getBytes()));
	}
}

File: app/src/main/java/com/sk/revisit/managers/WebStorageManager.kt
package com.sk.revisit.managers

import android.net.Uri
import android.webkit.URLUtil
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import com.sk.revisit.MyUtils
import com.sk.revisit.Revisit
import com.sk.revisit.log.FileLogger
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import okhttp3.Headers
import java.io.*

class WebStorageManager(private val utils: MyUtils) {

    val urlLogger: FileLogger
    var requestLogger: FileLogger
    val loggingExecutor: ExecutorService = Executors.newSingleThreadExecutor(LoggingThreadFactory())

    fun saveUrl(url: String){
        executorService.execute(()->{
            urlLogger.log(url)
        })
    }

    fun getResponse(request: WebResourceRequest): WebResourceResponse? {
        Revisit.requests.incrementAndGet()

        if (GET_METHOD != request.method) {
            return null
        }

        val uri = request.url
        val uriStr = uri.toString()
        saveUrl(uriStr)

        if (!URLUtil.isNetworkUrl(uriStr)) {
            return null
        }

        val localPath = utils.buildLocalPath(uri) ?: return null

        val localFile = File(localPath)
        val fileExists = localFile.exists()

        return if (fileExists) {
            if (Revisit.shouldUpdate && Revisit.isNetworkAvailable) {
                utils.download(uri, createDownloadListener(uriStr, localPath))
            }
            loadFromLocal(localFile, uri)
        } else {
            if (Revisit.isNetworkAvailable) {
                utils.download(uri, createDownloadListener(uriStr, localPath))
                loadFromLocal(localFile, uri)
            } else {
                saveReq(uriStr)
                createNoOfflineFileResponse()
            }
        }
    }

    private fun createDownloadListener(uriStr: String, localPath: String): MyUtils.DownloadListener {
        return object : MyUtils.DownloadListener {
            override fun onStart(uri: Uri, contentLength: Long) {
                // ToDo create a item_url_log
            }

            override fun onSuccess(file: File, headers: Headers) {
                Revisit.resolved.incrementAndGet()
            }

            override fun onProgress(p: Double) {
                // urlLog.pb.setProgress(p)
            }

            override fun onFailure(e: Exception) {
                Revisit.failed.incrementAndGet()
                saveReq(uriStr)
            }

            override fun onEnd(file: File) {
                // urllog.pb.visible=gone
            }
        }
    }

    private fun loadFromLocal(localFile: File, uri: Uri): WebResourceResponse? {
        val localFilePath:String = localFile.absolutePath
        return try {
            val mimeType:String = getMimeType(localFilePath, uri)
            Revisit.resolved.incrementAndGet()
            val inputStream = FileInputStream(localFile)
            val response = WebResourceResponse(mimeType, UTF_8, inputStream)
            response.responseHeaders = mapOf("Access-Control-Allow-Origin" to "*")
            response
        } catch (e: FileNotFoundException) {
            Revisit.failed.incrementAndGet()
            null
        }
    }

    private fun getMimeType(localFilePath: String, uri: Uri): String {
        var mimeType:String = utils.getMimeTypeFromMeta(localFilePath)
        if (mimeType == null) {
            utils.createMimeTypeMeta(uri)
            mimeType = utils.getMimeType(localFilePath)
        }
        return mimeType
    }

    private fun createNoOfflineFileResponse(): WebResourceResponse {
        return WebResourceResponse(
            DEFAULT_MIME,
            UTF_8,
            ByteArrayInputStream(NO_OFFLINE_FILE_MESSAGE.toByteArray())
        )
    }

    companion object {
        private const val DEFAULT_MIME = "text/html"
        private const val GET_METHOD = "GET"
        private const val UTF_8 = "UTF-8"
        private const val NO_OFFLINE_FILE_MESSAGE = "No offline file available."
        protected val TAG: String = this::class.java.simpleName
    }

    // Thread factory for logging tasks (lower priority)
    private class LoggingThreadFactory : ThreadFactory {
        fun newThread(r: Runnable): Thread {
            val t : Thread = Thread(r, "MyUtils-Logging-Thread")
            t.priority = Thread.MIN_PRIORITY
            return t
        }
    }
}
